presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxBy.java

```diff
@@ -95,40 +144,139 @@ private InternalAggregationFunction generateAggregation(Type valueType, Type key
 
     private static List<ParameterMetadata> createInputParameterMetadata(Type value, Type key)
     {
-        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, key), new ParameterMetadata(BLOCK_INDEX));
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INPUT_CHANNEL, key), new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void input(boolean min, Type keyType, MaxOrMinByState state, Block value, Block key, int position)
-    {
-        if ((state.getKey() == null) || state.getKey().isNull(0) ||
-                compare(keyType.compareTo(key, position, state.getKey(), 0), min)) {
-            state.setKey(key.getSingleValueBlock(position));
-            state.setValue(value.getSingleValueBlock(position));
+    private StateCompiler.StateClass generateStateClass(Type keyType, Type valueType, DynamicClassLoader classLoader) {
```

Comment:
It was hard to follow and debug, but I somehow enjoyed doing it. I guess one benefit (other than hopefully some performance gains which I probably should add a benchmark to test) would be this diff is net negative in lines of code, while implementing all possible interfaces is a lot of code. To me less code is almost always better.